evm/src/witness/operation.rs

```diff
@@ -162,6 +165,61 @@ pub(crate) fn generate_keccak_general<F: Field>(
     Ok(())
 }
 
+pub(crate) fn generate_poseidon_general<F: RichField>(
+    state: &mut GenerationState<F>,
+    mut row: CpuColumnsView<F>,
+) -> Result<(), ProgramError> {
+    row.op.poseidon_general = F::ONE;
+    let [(context, _), (segment, log_in1), (base_virt, log_in2), (len, log_in3)] =
+        stack_pop_with_log_and_fill::<4, _>(state, &mut row)?;
+    let len = u256_to_usize(len)?;
+
+    let base_address = MemoryAddress::new_u256s(context, segment, base_virt)?;
+    let input = (0..len)
+        .map(|i| {
+            let address = MemoryAddress {
+                virt: base_address.virt.saturating_add(i),
+                ..base_address
+            };
+            let val = state.memory.get(address);
+
+            val.0[0]
+        })
+        .collect_vec();
+    log::debug!("Hashing {:?}", input);
+
+    let mut padded_input = input.clone();
+    if padded_input.len() % POSEIDON_SPONGE_RATE == POSEIDON_SPONGE_RATE - 1 {
+        padded_input.push(1);
+    } else {
+        padded_input.push(1);
+        while (padded_input.len() + 1) % plonky2::hash::poseidon::SPONGE_RATE != 0 {
+            padded_input.push(0);
+        }
+        padded_input.push(1);
+    }
```

Comment:
We probably want padding to be done either directly on the kernel side, otherwise we would need 2 distinct instructions: 1 for hashing arbitrarily long data (leaves), and 1 for compressing 2 digests into 1 (i.e. without padding), which cannot happen here without an extra permutation.